### ISSUE TITLE
Spotlight: Use mouse wheel for navigation in place of 5-way

### ISSUE THREAD

**Issue by webOS101** [enactjs/agate#29]
Experiment with how apps could be controlled with a mouse wheel instead of 5-way.  What changes would be needed to spotlight?  To apps?  Would it require apps to specify navigation or could we figure out a consistent way to manage it without configuration?  Would new spotlight events be required?

**Comment by ryanjduffy** [comment]
Took a pass at this. Consider it WIP at this point.

https://github.com/enactjs/enact/tree/feature/spotlight-tab

**Comment by germboy** [comment]
I had some thoughts/questions after looking through this.

Ryan's initial approach looks 👍. I have only minor concerns. It looks like the tab key does generate a keyup event (while not generating a keypress event), so that can be used to reset the accelerator as usual instead of using a job. Without it, spotlight can fail to set focus in some cases when initially switching from pointer to 5way mode.

We need to clarify/define the `Scroller` behavior in agate. As we're implementing a wheel-driven 3-way behavior, how should that behave in a `Scroller`. Unlike moonstone, I'd imagine that using a scrollwheel 
in a `Scroller` should not scroll the viewport if we're using the scrollwheel to change focus. However, like moonstone, I'd think that we would use the "scroll into focus" behavior when using the scrollwheel to change focus in a `Scroller` and the focus moves outside the scroller. Would an agate `Scroller` use the ui `Scroller` prop `noScrollByDrag`, like moonstone. I'd think so, but there may be use-cases in agate I'm not aware of that may think otherwise.

Addition of scrollwheel support in spotlight would mean an adjustment may need to be made in ui and moonstone `Scrollable` support for scrollwheel. We'd need to make sure we don't cause regressions or unexpected behavior there. I have a poc solution for agate that uses the assumptions listed above (moonstone-like scroller using "scroll into focus" and a prop to not handle the wheel in `ui/Scrollable`)

I also investigated container support for tabindex, allowing containers to be prioritized as next targets. It's probably future work and may not need to be immediately implemented. However, it'd be nice to define how it could work. A container could perhaps define whether all (or only 1/some) of its Spottable descendants should be cycled next - or if there should be an order within them :tabception:

**Comment by germboy** [comment]
@webOS101 pointed out to me that RTL may pose a problem for tabindex ordering. I've tested general layouts for RTL (using ordered and unordered tabindex) across components and things are mostly ok, but it has uncovered some areas that need work. There are components with horizontal layouts (`IncrementSlider`) that work properly only in LTR and others (`Picker`) that only work properly in RTL. I'll take a look at normalizing this behavior.

**Comment by germboy** [comment]
I have added fixes for the concerns above in a working branch: https://github.com/enactjs/enact/tree/feature/spotlight-tab-jeremy

The LTR/RTL issue arose from the way spottble components are ordered in the dom. Using a `querySelector` to retrieve a list of spottable components will give us a list of items in the order they appear in the dom. So when moving to the prev/next spottable in a component (`Picker` for example), we're expecting to navigate in layout order (decrementer <> incrementer). However, this component's incrementer was the first in the dom and repositioned as visibly last (via css). The reverse is true for the `IncrementSlider` component. If we make these components locale-aware, we can render the dom in the order the browser will naturally expect when creating our list of spottable components.

A more complex component may need to reposition its spottable controls in a layout that can not be expected (`VirtualList`, for example). That component would then be responsible for handling the wheel events to focus to the next item and preventing propagation to spotlight.

I experimented with using a secondary criterea when sorting elements (`tabindex` then a `data-index` attribute value, matching the `VirtualList` method for specifying expected order) that could possibly be added to `Spottable` as a prop). The hope was to have an option to bypass forced layouts, but this didn't seem doable since we couldn't be sure of the owner of the elements (since they may not use containers) and wouldn't know the amount of elements to order for the 2nd criterea). This would also introduce multiple tab orders which could be confusing.
All-in-all, the it seems reasonable to me to lay out elements in the way the browser would expect them, but I'd love to hear any thoughts/input.

I also lightly experimented with container ordering via tabindex, which seems doable as a means to direct focus to containers over elements, but would need a bit of refinement and direction (:rimshot:) since there are no behaviors defined for that and it could mean a number of different focus change options).

**Additional Considerations**
1) `Scrollable` and `VirtualList` handle the wheel event as a way to scroll the viewport. This conflicts with how spotlight would use the wheel event to scroll to the next/previous spottable. I'd expect the current scroll behavior to remain, so I'm preventing wheel event propagation in `Scrollable` and `VirtualList` until we can further define the behavioral conflict and what we may want to change.
